pkgs/build-support/rust/default.nix

```diff
@@ -89,7 +90,9 @@ stdenv.mkDerivation (args // {
   patchRegistryDeps = ./patch-registry-deps;
 
   nativeBuildInputs = nativeBuildInputs ++ [ cacert git cargo rustc ];
-  buildInputs = buildInputs ++ stdenv.lib.optional stdenv.hostPlatform.isMinGW windows.pthreads;
+  buildInputs = buildInputs
+  ++ stdenv.lib.optional stdenv.hostPlatform.isMinGW windows.pthreads
+  ++ stdenv.lib.optional stdenv.hostPlatform.isDarwin darwin.apple_sdk.frameworks.Security;
```

Comment:
Is it needed for the stdlib? @LnL7 usually prefers to have impure macOS dependencies on a per-package base.

Reply:
Sorry I don't understand, why is it acceptable for `go` and not for `rust`?

Reply:
I think in go is required by its standard library, but in rust is only specific crates which depend on Security. e.g. `getrandom` crate in shadowenv

Reply:
https://github.com/NixOS/nixpkgs/blob/19a0b38cbe352f964acafb139e2a37dd082c9e3d/pkgs/development/compilers/rust/rustc.nix#L132-L134

https://github.com/NixOS/nixpkgs/blob/19a0b38cbe352f964acafb139e2a37dd082c9e3d/pkgs/development/compilers/rust/cargo.nix#L21-L22

Reply:
Rust uses external crates, including `rand`:

https://github.com/rust-lang/rust/blob/9d38dc22e5d543b9c350dd48ac014135547e7953/Cargo.lock#L2478

So, it's not surprising that rustc and cargo require `Security`.

I don't think the standard library depends on `Security`. E.g. the `skim` derivation does not have `Security` in `buildInputs` and builds fine on Darwin.

Reply:
I guess my point is that the end result isn't really any different to `go`.

`rustc` / `cargo` need it and ~40% of `buildRustPackage`s also need it.

Reply:
I don't really like this, Security framework is very similar to openssl/curl. It's not because many projects depend on these libraries that we should include them in the standard environment for convenience.

I'm not all that familiar with go but if I recall correctly the big difference there is that the http library is part of the stdlib there which means ssl and this Security are a builtin dependency of the runtime.

Reply:
cc also @matthewbauer who maintains a lot of darwin stuff.

Reply:
I think we're kinda past the point of including it for convience, by not including it we're just adding cruft to packages and making more work for ourselves.

I'm pretty sure if I keep looking I'll be able to get `Security` past 50% for `buildRustPackage`s.

Reply:
Well the real problem here is that we don't build crates separately, this is an issue for all native libraries. With something like crate2nix only a few special crates would have to be annotated at which point all other crates can inherit that automatically.

Reply:
Not to import the discussion of #89563 here. But the underlying problem here is that `buildRustPackage` et al. do not properly define the dependency graph in Nix, so it is all or nothing. If dependent crates were also defined as Nix derivations (as e.g. `buildRustCrate` does), then we could just add the `Security` dependency to the small number of crates (`rand`? etc.) that require it.

Adding `Security` to the `buildInputs` of all `buildRustPackage` derivations because (a large) portion of them require it, seems like yet another step away from the strengths/principles of Nix (properly defined dependency graphs).

Reply:
Yes, `buildRustPackage` already has multiple issues. I don't see why `Security` is a hill to die on.

Reply:
So solutions like `crate2nix` mean that the darwin `Security`, etc inputs in packages is literally cruft?

We're adding it to packages knowing that it will be removed anyway?

Reply:
> Yes, `buildRustPackage` already has multiple issues. I don't see why `Security` is a hill to die on.

I agree that it is very tedious, so perhaps we should be pragmatic about this particular case. It is not required by the Rust standard library, but that's because the standard library does not carry (proper) RNG support and most packages do need it anyway (somewhere).

Reply:
I'd like to continue with this but I don't know that I've got the time to see it through at the moment.

Apologies for my ranty comments, rereading them I probably come across harsher than I intended.